libzecale/circuits/fields/fp12_2over3over2_gadgets.hpp

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) 2015-2020 Clearmatics Technologies Ltd
+//
+// SPDX-License-Identifier: LGPL-3.0+
+
+#ifndef ZECALE_CIRCUITS_FIELDS_FP12_2OVER3OVER2_GADGETS_HPP__
+#define ZECALE_CIRCUITS_FIELDS_FP12_2OVER3OVER2_GADGETS_HPP__
+
+#include "libzecale/circuits/fields/fp6_3over2_gadgets.hpp"
+
+namespace libzecale
+{
+
+template<typename Fp12T>
+class Fp12_2over3over2_variable : public libsnark::gadget<typename Fp12T::my_Fp>
+{
+public:
+    using FieldT = typename Fp12T::my_Fp;
+    using Fp6T = typename Fp12T::my_Fp6;
+
+    Fp6_3over2_variable<Fp6T> _c0;
+    Fp6_3over2_variable<Fp6T> _c1;
+
+    Fp12_2over3over2_variable(
+        libsnark::protoboard<FieldT> &pb, const std::string &annotation_prefix);
+    Fp12_2over3over2_variable(
+        libsnark::protoboard<FieldT> &pb,
+        const Fp12T &v,
+        const std::string &annotation_prefix);
+
+    Fp12_2over3over2_variable(
+        libsnark::protoboard<FieldT> &pb,
+        const Fp6_3over2_variable<Fp6T> &c0,
+        const Fp6_3over2_variable<Fp6T> &c1,
+        const std::string &annotation_prefix);
+
+    Fp12T get_element() const;
+    void generate_r1cs_witness(const Fp12T &v);
+};
+
+/// Follows implementation in libff::Fp12_2over3over2_model, which is based on
+/// Devegili OhEig Scott Dahab "Multiplication and Squaring on Pairing-Friendly
+/// Fields"; Section 3".
+///
+/// As an order 2 extension of Fp6, Fp12 has elements of the form
+///   (a0, a1) = a0 + w.a1 for a0, a1 in Fp6.
```

Comment:
`w.a1` -> `w * a1` ?
(likewise for the lines below)

Reply:
switched to `*` for "multiplication"